### ISSUE TITLE
Design

### ISSUE THREAD

**Issue by klemens-morgenstern** [mw-sc/mw.coroutine#1]
# Preface

This are the design notes for my coroutine library - @timonwalther  @odinthenerd @CvRXX  feel free to participate.

First off: my [CI with qemu](https://travis-ci.org/mw-sc/mw.test.example) works, including [coverage](https://coveralls.io/github/mw-sc/mw.test.example), which means I can build a properly tested library now.

# Structure 

Given that there are such awesome things as threads, coroutines, asio, fiber we need a __context__ library, which in the future can be used for threading and what-have-you. I'll look up `boost.context` which has an ARM implementation and FreeRTOS - I haven't looked up yet if I can shorten the code, when exceptions are disabled. 

# Basic design

The basic design should be as simple as possible, just a coroutine and it should take two pointers for the stack. In addition it will have a function that tells you if you're still in the stack or not + plus a few bytes reserved. Thirdly it shall have an error handler, which could for example throw an exception.

What is also required is a way to chancel the coroutine. Since the stack can be chosen randomly, it is possible to put a piece of protected memory before it (sp is normally descending on arm) and thus an interrupt can occur if the coroutine oversteps the stack. If that happens it must be possible to chancel the coroutine. If we have that situation but no exceptions, we need a **manual unwind** function --> I have to do some research.

## Deep Sleep

If we have coroutines we can also store them for a deep sleep. This also works better when combined with a unwind/chancel function, because some coroutines might not be storable during deep sleep. 

## Fiber

Now another application, or way to have concurrency on controllers is using interrupts. And here the sequence of interrupts can be somewhat complex, e.g. you can setup an SPI purely by enabling the interrupt and saying it should init and do all the other steps through the interrupt. That is what is typically done in a state-machine, but it doesn't have to. A better model would actually be a coroutine if we can make it work, including error handling. Maybe this should better be part of a `fiber` library, but let's take a example, where I have an `init` function, called from main, and then several steps in the coroutine to initialize the hardware and walk through that. *Those are just thoughts, please comment with better ideas!*

### Contrived example 

Alright, so we have a hardware, with an interrupt, and we want to init that, and do three operations. The state machine would look like this:

```cpp
state_machine_data hw_sm;

void hw_IRQ()
{
   switch (hw_status())
   {
     case hw_inited:
             hw_sm.set_inited();
             hw_operate();
             break;
     case hw_success:
             hw_sm.success();
             if (hw_sm.complete())
                 hw_shutdown();
             else
                 hw_operate();
             break;
     case hw_error:
             hw_sm.error(hw_get_error());
             hw_shutdown();
   }
}

int main() {hw_init();}
```

Now what I'd like to do is something like that (pseudo-code, of course):

```
error_type hw_cr()
{
    //this is called by main
    struct on_scope_exit_t 
    {
       ~on_scope_exit() {hw_shutdown();}
    } on_scope_exit;
   if (yield() != hw_inited)
       return hw_error();
   
   for (auto i = 0; i < 3; i++) //operate it three times
   {
         hw_operate();
         if (yield() != hw_success)
             return hw_error();
   } 
}

int main()
{
    hw_coroutine cr ( &hw_cr ) ;
    future<error_type> ft = cr.run();

   return 42;
}
```

Well something like that, I'm unsure if that should be part of a *coroutine* or *fiber* library, but it sure seems like the right model to me.

**Comment by klemens-morgenstern** [comment]
New Idea: if I can use a fiber/coroutine to get something through the DMA onto the stack, I'll die a happy man.

```cpp
std::array<char, 256> buf;
yield_dma(buf); //let the dma do it's magic and only reenter once it's done.
```

**Comment by klemens-morgenstern** [comment]
New Idea: allow repassing the arguments on entry, that also takes care of the unwinding. I.e.:

```
coroutine<bool(bool)> cr = 
  [](bool stop){
      yield true;
      if (stop) 
         return false;
      yield true;
      if (stop)
          return false;
     return true;
  };
```

Alternatively though, `yield` could carry a value, since I might just want to pass one value in, while the arguments should stay the same.

**Comment by klemens-morgenstern** [comment]
Note: consider how coroutines could be forked --> possible if the stack-container can be copied. If using the heap something is needed as well.

**Comment by klemens-morgenstern** [comment]
Question: does it need exception handling?

**Comment by odinthenerd** [comment]
not for any of my designs at the moment but it think exception handling may become more viable in the domin in the future with better tooling

**Comment by klemens-morgenstern** [comment]
Yeah, I've been thinking some more, and I think it's a bad idea. You can use [std::exception_ptr](http://en.cppreference.com/w/cpp/error/exception_ptr) to get it out (which I would do internally anyway) and I'll add an example + test for this.
